### DISCUSSION TITLE
4.11.0-RC binaries available

### DISCUSSION THREAD

**Discussion by paulpv** [DistroAV/DistroAV#853] [General]
I went ahead and tagged the PR to generate an official build for everyone to test out.

Please try to build at...
https://github.com/Palakis/obs-ndi/actions/runs/4286792668

...and let me know if you have any problems.

**Comment by truthseeker308** [comment]
Just testing(can't test for the audio).  Video works fine(Birddog P200 camera), but noticing that green(preview) tally doesn't light up when camera is in Preview in OBS Studio.  Latest versions of NDI 5.5, 4.11.0-RC and OBS 29.0.2.

**Reply by paulpv** [reply]
Thanks for testing and reporting.

Noted.

I've never played with or used Tally. I'll look in to it.

**Comment by KattPhloxworthy** [comment]
One thing I'm noticing is a VERY long delay between video and audio if I use OBS Studio on a machine to output a full program-grade NDI stream to another computer, which has an ingester source scene, which is then streamed to somewhere like Twitch.  Encoding at 720p59.94 (don't give me static about my "non-standard framerate; I don't want to hear about it) at 4Mbps to Twitch, I see a one-second delay between the audio and the video, the audio lagging behind the video.  I don't think any manner of offset is going to help with this.

Any suggestions with this?  Or is this a peculiarity with NDI?

In the meantime, I'm sticking with Teleport, which I am having far better luck with for the intermachine link between the "studio" and "encoder" machines on my network.

And some specifics about my setups:

Studio machine:
AMD Ryzen 7 5800+, 64GB RAM
Windows 10 Pro 21H2

Encoder machine:
Core i5 6500, 32GB RAM
Windows 10 Pro 22H2

Both systems:
OBS Studio 29.0.2
NDI Tools 5.5.3
OBS NDI plugin 4.11.0-RC.

The "studio PC" has Tools/NDItm Output Settings set to the following:

Main Output:  Checked
Main Output name:  OBS Send
Preview Output:  Unchecked
Preview Output name:  (Grayed out, but currently says "OBS Preview"; I do not use a preview stream)
Tally:  Program and Preview - unchecked

The "encoder PC" has a scene with the NDI source set to ingest the previous PC's NDI stream with the following settings:

Source name:  (Source of my "studio PC")
Bandwidth:  Highest
Sync:  I've tried both "Source timing" and "Network" to no avail.
Allow hardware acceleration:  I've tried both checked and unchecked to no avail
Fix alpha blending (adds a filter to this source):  Unchecked
YUV Range:  Partial
YUV Color Space:  BT.709
Latency Mode:   I've tried both "Normal (Safe)" and "Low (Experimental)"
Enable Audio:  Checked

An example of this delay can be seen here at about the 4:46 mark where there's a countdown.

https://www.twitch.tv/videos/1754214364

I still do use NDI to project a completed scene of my VTuber with an overlaid "throwing program" from my "VTuber PC" to my "studio PC" because I want/need the alpha channel support.  Between my "studio PC" and my "encoder PC", I have to use Teleport, which works, but I'd rather be standardized throughout my streaming operation.

If there is anything I should be trying or am missing, I'm all ears.  Likewise, if there is anything additional needed, please let me know.

**Reply by paulpv** [reply]
Thanks for testing and reporting.

Noted. The 4.11.0-RC is basically the old original plugin updated for OBS28+/QT6/64bit-only and a few other small patches.
tt2468 started his `rewrite` branch almost a year ago with one goal of being a lot more performant/efficient.
The old code could probably use some threading improvement learned from the `rewrite` code.

I will keep an eye/ear out for this and look in to it if I see it affects everyone.
If it only affects a small percentage of users then I may wait to look in to this until after I get 4.11.0 released.

**Reply by KattPhloxworthy** [reply]
I've done some further extensive testing with not just my current "studio PC" which is also my gaming PC, but a soon-to-be-pressed-into-service "new studio PC", dedicated solely to compositing my streams.

Here's what I've been able to find:

Gaming PC OBS Studio NDI Output Settings Main Program send:  Audio has a one second delay on full stream.  Can compensate by receiving separate audio and video streams, then offsetting the video by that one second using the Video Delay (Async) plugin on audio on the video stream.  It's not totally critical since there's already a delay introduced by Twitch's RTMP network and this delay only happens between my gaming PC and the encoder PC.  While it will work, I consider this as less-than-ideal.  However, if sticking with NDI isn't critical (which it really isn't; it's just that I'd rather standardize on NDI throughout my operations), I can use the Teleport plugin instead, and it will work properly.

Gaming PC with NewTek's screen capture and audio capturing a video player:  Audio and video arrive at any NDI receiver without any issues.

New studio PC OBS Studio NDI Output Settings Main Program send:  No delay at all.  No workarounds needed.  I can use a single NDI source on the encoder PC.

The specs on the "new studio PC" are as follows:

Hewlett Packard EliteDesk 800 G2
Core i7 6700
32GB RAM
AMD Radeon RX550 video card

All this said, it does look like a very isolated "special case" scenario you describe above.  It also does bear noting that I have a ton of plugins in that OBS Studio setup that's experiencing the delay out the back of OBS Studio via NDI.

**Reply by paulpv** [reply]
@KattPhloxworthy Fairly odd behavior that `Gaming PC` is showing a ~1 second delay, but it is good at least that the `Studio PC` has no delay.

I have seen the occasional delay when dealing w/ my code changes, but I think it was usually because I had some old NDI libraries being loaded, or even OBS was loading an old obs-ndi version.  
I cleaned up searched my whole system for `*ndi*.*` and made sure everything was uninstalled before I re-installed the version I am working on.
Uninstalling and reinstalling OBS and/or NDI also seemed to help.
All in all, I accounted for all of my experiences just being the result of working w/ 20+ different compiled versions throughout the day (I sometimes intentionally went back to the old version to compare/contrast) and getting some of them cross-pollenated.

Let me know if you find the source of your `Gaming PC`'s ~1 second audio delay.

**Reply by DPB72** [reply]
I am using obs-ndi 4.11.1, and am also experiencing this 1 second audio delay on the NDI output by the plugin. I get this even in NDI Studio Monitor running on the *same computer* as the OBS generating the NDI. I have two video files that I recorded simultaneously -- one was recorded in OBS, and the other was recorded in NDI Studio Monitor: in the audio is the former is fine (in sync and audible), whereas the audio of the latter is both out of sync (the audio being around a second behind the video) and barely audible.

I've been using OBS for several years to stream to YouTube (successfully), but I'm now trying to get OBS also to output NDI for a TV relay (in the same building) to pick up on the LAN, but finding the NDI generated is not usable (as described above). Any suggestions very much appreciated.

**Comment by Zacharybinx34** [comment]
Running on a Mac M1, and when I start OBS it says "NDI Runtime not found" and the link to get it doesn't work. 

Any ideas? I just installed the latest for MAC. 

Edit: Installing this fixed it, but I'm wondering why it's even needed? 
NDI runtime: [ndi-runtime-5.0.10.1-macOS.pkg](https://ndi.palakis.fr/runtime/ndi-runtime-5.0.10.1-macOS.pkg)

**Reply by paulpv** [reply]
Everyone has to separately install the NDI runtime or tools. We are not allowed to distribute them in our installer.

We could look in to improving the installer to tell the user to install the NDI runtime/tools.

**Reply by paulpv** [reply]
At a minimum I should update the readme with a more end-user friendly `Installation` section that describes the installation steps and that tries to make it obvious that the NDI runtime/tools need to be separately installed.

I have one tweak to make to the Linux build+install process, so I make this change when I do that.

**Comment by beeradmoore** [comment]
EDIT: Solved. Firewall issue.

Not able to get this working. New to it all so unsure if its this build or my setup.

My setup is to show an Application on my Mac Mini on my Windows PC which is running OBS. Mac Mini which is running NDI Tools 5.5.3.16. I did install obs-ndi-4.11 but from what I understand it is not necessary. Windows 10 PC has OBS 29.0.2. NDI Tools 5.5.3.0 and obs-ndi-4.11.

I run Scan Converter (from NDI Tools) on the Mac Mini and select the application to share. On Windows I test with NDI Studio Monitor, I can confirm that the NDI signal is getting through the network from the Mac to the Windows machine.

I close that, then in OBS I add an NDI source into my scene, however the Mac does not appear in the source's dropdown. 

Is this a bug with the plugin, an issue with how I set things up, or just not how this plugin is intended to work?

**Reply by paulpv** [reply]
If you just want to view the Mac desktop on the Windows PC then you do not need OBS or this plugin on the Mac.

For years I have seen obs-ndi sometimes unable to detect NDI streams.

I can often resolve the problem by rebooting all involved computers and/or my network hardware.

A few times I had been unable to resolve the problem and gave up for a few days, only to try again a few days later and things worked fine. /shrug/

Try using the NDI Video Monitor on the Mac to see if it can see the NDI Screen Capture (I think they call it Scan Converter). You should be able to.

Try another PC?

Are the Windows PC and Mac on the exact same network? Is one of them wireless and the other wired? Try making them both wired.

**Reply by beeradmoore** [reply]
Whops, I forgot to include why I use OBS in the first place with this setup.  I use virtual cam for work. When I go to share my screen (to show things from the Mac) I change scenes in OBS to bring that up. For a while I was using a capture card but it often played funny buggers with resolution and display refresh rates on the source computer so I was looking to this as an alternative. 

Same issues after a reboot.

Using NDI Video Monitor on the Mac and using NDI Screen Capture on the Windows PC I am able to connect and  view the other display correctly.
Using NDI Video Monitor on the Windows PC and using NDI Scan Converter on the Mac I am able to connect and view the other display correctly.

As far as I can tell with that test NDI over the network is working fine. It's just not detecting and using it in OBS (or OBSs networking).

Both are hard wired directly into the router sitting between them. 

BUT
Installing NDI on a second PC (Windows 11,  also hard wired into the same network), rebooting, install OBS plugin, open OBS and both computers (original Windows 10 PC and Mac Mini) show up perfectly fine.

I'll check if OBS is firewall blocked on this first PC, other than that I wonder if it could be a Win10 vs Win11 issue but that seems unlikely.

**Reply by beeradmoore** [reply]
Checked firewall rules on Win11 PC and I could see OBS there.
Checked firewall rules on Win10 (first) PC and there was no OBS. 

Added OBS, restarted OBS, everything worked as expected.

That explains why NDI was working for NDIs own tools as it would have been allowed in firewall but OBS for some reason never prompted for firewall access. Thanks for helping me keep digging and track that down. Glad it wasn't the plugin after all!

**Reply by paulpv** [reply]
Fantastic news!

Yes, this plugin seems a little finicky.

Even **before I got involved in this project** I got the feeling that this plugin seems to work >95% of the time and that the "few" [200+] [Issues](https://github.com/obs-ndi/obs-ndi/issues) are...despite most of them probably being knowledgable users that really do know what they are doing...exceptions.

The trick is going to be finding which bugs are legit bugs in the code, and which bugs are just legit bugs in the users' setups.

Your report helps. Thanks! Firewall is definitely a complication. Troubleshooting it with a 3rd PC usually helps to spotlight that as the difference. Now if only all of the other [200+] [Issues](https://github.com/obs-ndi/obs-ndi/issues) were as east to resolve! :)

Take care!

**Comment by jbcurler2010** [comment]
I was always told to ensure that network was set to Private on all devices
and make sure Firewall is off for Private networks.

John Benton
Curling Stadium
612-202-4588

**Reply by paulpv** [reply]
I agree with the first recommendation to ensure that your local/home network is set to Private.
I disagree with the second recommendation to turn the firewall off for Private networks.
Sure, that could simplify much of the connectivity problems, which may be fine during proof-of-concept, but even a Private network should have some sort of strong firewall protection in place with judicious selective application or port exceptions.

**Comment by Diversq** [comment]
I installed the RC. Unfortunately it did not correct the problem.

Im running a Windows 10 PC with the NDI ScreenCapture HX and picking up the stream on a 2020 Mac M1.

If I use the NDI Video Monitor the video and audio is perfect.
If I use the OBS plugin the Video is fine and when monitoring the audio in OBS thats fine too. But when Streaming or Recording I get crackle and static. Some audio but not usable quality.

This is a youtube link of a recording to show what I mean.

[Link to youtube vid](https://youtu.be/K-bVBN3JJxI)

**Reply by paulpv** [reply]
I watched your vid and was going to comment there, but might as well here instead.

My first impression of the audio, listening in headphones, was that it sounds like variable a pitch and slowing down problem on both left and right channels, as if the CPU cannot keep up, which is different from the consistent right channel audio consistently clipping.

I downloaded the video (w/ yt-dlp) and opened the audio in audacity and saw what I thought:
![image](https://user-images.githubusercontent.com/1393897/223280809-bdbdb132-92a5-4e96-93ae-c9951e6af587.png)
(12 obvious glitches in 1.5 seconds)

Gaps in both the left and right channel audio; not the same known bug that was fixed.

I use a MacBook Pro 2021 M1 from time to time, and mine seems to also **sometimes** have a hard time keeping up **transmitting** audio/video, but I have not seen it unable to keep up **receiving** audio/video..  
When transmitting, a few times I observed the video and audio drifting slowly further and further behind.  
I thought I had fixed that problem or at least accounted for it and hoped that it was no longer an issue.  
But at no time to I recall the audio slowdown sounding like that.

Thinking about it...

**Reply by paulpv** [reply]
The issue I mentioned at the end of my previous comment, about audio/video drifting slower and slower, was fixed by this change:

https://github.com/obs-ndi/obs-ndi/commit/36c5124b68cf89e597eb8211fdea6ec3c2a625b4#diff-a8d90a655e8e0ec80bcc9891814e7139493615e5586601db725e102e15073186R294-R295
<img width="1307" alt="image" src="https://user-images.githubusercontent.com/1393897/223282095-a6ee5fbb-530a-45ab-a8aa-9dedc0de1199.png">

One flaw I know about that change, and even comment an allusion to, and is even documented in the NDI SDK docs, is that there is only one video buffer being used, and it is being handed directly to the `send_send_video_async_v2` method.
The NDI SDK docs say:
<img width="1131" alt="image" src="https://user-images.githubusercontent.com/1393897/223282870-bb676dbc-a4e7-4089-a81e-70117a4fa247.png">
I know about this and plan to improve it in a future version.
That said...this is all about video, which isn't really being complained about [yet].

There is no equivalent `send_send_audio_async_v3`; we are just using `send_send_audio_v3`.
But there are a few other things that can be done to improve performance and/or quality a bit, mostly related to timestamps and complicates A/V synchronization (which there may already be some options in the existing plugin, but I'll have to investigate that more).

All of this is related to NDI Output.
None of this is related to NDI Input, which is what you say you are observing @Diversq.

Still, there is an equivalent input/RX paradigm that could perhaps be made more efficient.
A 2020 Mac M1 may not be powerful enough with the existing code.
If you could try on a more powerful Mac that would give a data point, but that would mostly be for your benefit; as I mentioned already I run a MacBook Pro 2021 M1 and don't see this problem.

**Comment by DC47561** [comment]
Hi.

I'm running Ventura 13.2.1, OBS 29.0.2 (ARM version) and NDI version 4.11.0-RC on my machine.

I'm looking to get my setup going without using Rosetta but cannot find a way to install NDI Redistributable V5.

The error I get is "NDI Runtime not found. - Download the installer here: http://new.tk/NDIRedistV5Apple."

Clicking on the link results in the page failing to load. I get "This site can't be reached" 

I think the website may be down.


Many thanks.

**Reply by paulpv** [reply]
Yes, NewTek NDI has coincidentally unfortunately been moving to new URLs and their download links have been misbehaving for almost a week.
They are supposedly aware of and working on it.
https://forums.newtek.com/threads/http-new-tk-seems-down-and-thus-http-new-tk-ndiredistv5-http-new-tk-ndiredistv5apple.167380/
